lib/common/iso8601.c

```diff
@@ -1027,53 +1027,58 @@ parse_date(const char *date_str)
 }
 
 // Parse an ISO 8601 numeric value and return number of characters consumed
-// @TODO This cannot handle >INT_MAX int values
-// @TODO Fractions appear to be not working
-// @TODO Error out on invalid specifications
 static int
-parse_int(const char *str, int field_width, int upper_bound, int *result)
+parse_int(const char *str, int *result)
 {
-    int lpc = 0;
-    int offset = 0;
-    int intermediate = 0;
-    gboolean fraction = FALSE;
-    gboolean negate = FALSE;
+    unsigned int lpc;
+    int offset = (str[0] == 'T')? 1 : 0;
+    bool fraction = false;
+    bool negate = false;
 
     *result = 0;
     if (*str == '\0') {
         return 0;
     }
 
-    if (str[offset] == 'T') {
-        offset++;
-    }
+    // @TODO This cannot handle combinations of these characters
+    switch (str[offset]) {
+        case '.':
+        case ',':
+            fraction = true;
+            offset++;
+            break;
+
+        case '-':
+            negate = true;
+            offset++;
+            break;
 
-    if (str[offset] == '.' || str[offset] == ',') {
-        fraction = TRUE;
-        field_width = -1;
-        offset++;
-    } else if (str[offset] == '-') {
-        negate = TRUE;
-        offset++;
-    } else if (str[offset] == '+' || str[offset] == ':') {
-        offset++;
+        case '+':
+        case ':':
+            offset++;
+            break;
+
+        default:
+            break;
     }
 
-    for (; (fraction || lpc < field_width) && isdigit((int)str[offset]); lpc++) {
+    for (lpc = 0; (fraction || (lpc < 10)) && isdigit(str[offset]); lpc++) {
+        const int digit = str[offset++] - '0';
+
         if (fraction) {
-            intermediate = (str[offset] - '0') / (10 ^ lpc);
+            /* @TODO Previously, the code here essentially used:
+             *
+             *     *result += digit / (10 ^ lpc);
+             *
+             * With integer division, lpc > 0 will always yield a zero
+             * expression, and it should be 10 ^ (lpc + 1) anyway. Rather than
+             * fix it, we really just need to replace our ISO 8601 code with
+             * GDateTime.
```

Comment:
Not to mention that's a bitwise XOR rather than exponentiation

Reply:
lol, I completely overlooked that

Reply:
updated comment